frontend/src/metabase/query_builder/components/filters/pickers/DatePicker/DatePicker.unit.spec.tsx

```diff
@@ -0,0 +1,409 @@
+import React, { useState } from "react";
+import { render, screen, fireEvent } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+
+import { ORDERS } from "__support__/sample_database_fixture";
+import Filter from "metabase-lib/lib/queries/structured/Filter";
+
+import DatePicker from "./DatePicker";
+
+const ordersQuery = ORDERS.query();
+
+// this component does not manage its own filter state, so we need a wrapper to test
+// any state updates because the component's behavior is based on the filter state
+const DatePickerStateWrapper = ({
+  filter,
+  onCommit = jest.fn(),
+  onChange = jest.fn(),
+}: {
+  filter: Filter;
+  onCommit?: (arg: any) => void;
+  onChange?: (arg: any) => void;
+}) => {
+  const [filterValue, setFilterValue] = useState(filter);
+  return (
+    <DatePicker
+      filter={filterValue}
+      onFilterChange={(arg: any) => {
+        setFilterValue(arg);
+        onChange(arg);
+      }}
+      onCommit={onCommit}
+    />
+  );
+};
+
+describe("DatePicker", () => {
+  beforeAll(() => {
+    // this should keep these tests from behaving differently when run in the future
+    jest.useFakeTimers();
+    jest.setSystemTime(new Date("2020-05-01 08:00:00"));
+  });
+
+  afterAll(() => {
+    jest.useRealTimers();
+  });
+
+  const dummyFunction = jest.fn();
+
+  describe("Static States", () => {
+    it("renders a date picker component", () => {
+      const filter = new Filter(
+        [null, ["field", ORDERS.CREATED_AT.id, null]],
+        null,
+        ordersQuery,
+      );
+
+      render(
+        <DatePicker
+          filter={filter}
+          onFilterChange={dummyFunction}
+          onCommit={dummyFunction}
+        />,
+      );
+
+      screen.getByTestId("date-picker");
+    });
+
+    it("shows a date shortcut picker for an empty filter", () => {
+      const filter = new Filter(
+        [null, ["field", ORDERS.CREATED_AT.id, null]],
+        null,
+        ordersQuery,
+      );
+
+      render(
+        <DatePicker
+          filter={filter}
+          onFilterChange={dummyFunction}
+          onCommit={dummyFunction}
+        />,
+      );
+
+      screen.getByTestId("date-picker-shortcuts");
+    });
+
+    describe("Specific Dates", () => {
+      const singleDateOperators = ["=", "<", ">"];
+
+      singleDateOperators.forEach(operator => {
+        it(`shows a single specific date picker for a ${operator} operator`, () => {
+          const filter = new Filter(
+            [operator, ["field", ORDERS.CREATED_AT.id, null], "2020-01-01"],
+            null,
+            ordersQuery,
+          );
```

Comment:
This is entirely subjective, but WDYT about adding a helper to generate filters for tests? On the one hand, it's nice to see the entire filter passed, on the other hand, you need to spend time "parsing" the whole block to figure out what's different about each specific filter. We could have a utility setting some defaults and then each test would focus on overwriting values that are important for each specific test. For example, I see most of the tests if not all of them always pass a filter, then `null` and then `ordersQuery`. Those two arguments can be moved out. Again, this is purely subjective

Reply:
Also, you should be able to use `ORDERS.CREATED_AT.reference()` instead of duplicating `[ "field", ... ]` over and over again. Should we even extract it to constant somewhere at the top if it's used everywhere? For instance:

```js
// Somewhere at the top of the file
const CREATED_AT = ORDERS.CREATED_AT.reference()

// Example test

const filter = new Filter([ "=", CREATED_AT ])
```

Reply:
> This is entirely subjective, but WDYT about adding a helper to generate filters for tests? On the one hand, it's nice to see the entire filter passed, on the other hand, you need to spend time "parsing" the whole block to figure out what's different about each specific filter. We could have a utility setting some defaults and then each test would focus on overwriting values that are important for each specific test. For example, I see most of the tests if not all of them always pass a filter, then `null` and then `ordersQuery`. Those two arguments can be moved out. Again, this is purely subjective

I like the idea of helpers, but I think sometimes they can be more cognitive load than just writing things out explicitly in tests because it can be a little challenging to understand what all the helpers do on top of what all the tests do.

Nonetheless, your point is well-taken and I see a few opportunities to reduce some code duplicate here. 👍